### PR TITLE
the discount service api implementation

### PR DESCRIPTION
**Pull request by kyribamstraining**

### CODE REVIEW

File: discount/src/main/java/com/kyriba/discount/web/controller/DiscountController.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
+package com.kyriba.discount.web.controller;
+
+import com.kyriba.discount.domain.DiscountType;
+import com.kyriba.discount.domain.dto.DiscountDto;
```
Comment (ilya-a87) [inline discount/src/main/java/com/kyriba/discount/web/controller/DiscountController.java]: domain objects in API

Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/discount/web/controller/DiscountController.java]: I don't see any reason to add more objects right now. If I need an object to manipulate domain logic, then I will add this one in the future.

File: discount/src/main/asciidoc/index.adoc [hunk 2]
```diff
@@ -0,0 +1,89 @@
+= Discount microservice
+Andy Wilkinson; Jay Bryant
```
Comment (ilya-a87) [inline discount/src/main/asciidoc/index.adoc]: Andy Wilkinson; Jay Bryant?

Comment (kyribamstraining) [reply discount/src/main/asciidoc/index.adoc]: You are looking at the old version. I have already deleted this copy-paste

File: discount/build.gradle [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	mavenCentral()
+}
+
+ext {
+	set('snippetsDir', file('build/generated-snippets'))
+}
+
+dependencies {
+	implementation 'org.springframework.boot:spring-boot-starter-actuator'
+	implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
```
Comment (xpinjection) [inline discount/build.gradle]: We don't need JPA at that stage.

File: discount/src/main/java/com/kyriba/conference/discount/DiscountApplication.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.annotation.Bean;
+import springfox.documentation.builders.RequestHandlerSelectors;
+import springfox.documentation.spi.DocumentationType;
+import springfox.documentation.spring.web.plugins.Docket;
+import springfox.documentation.swagger2.annotations.EnableSwagger2;
+
+@SpringBootApplication
+@EnableSwagger2
```
Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/DiscountApplication.java]: Move to dedicated Swagger configuration in api package.

File: discount/src/main/java/com/kyriba/conference/discount/DiscountApplication.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+	public static void main(String[] args) {
+		SpringApplication.run(DiscountApplication.class, args);
+	}
+
+	@Bean
+	public Docket discountApi() {
+		return new Docket(DocumentationType.SWAGGER_2)
+				.select()
+				.apis(RequestHandlerSelectors.basePackage("com.kyriba.conference.discount.api"))
```
Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/DiscountApplication.java]: Add global API info and apply filter on URL level.

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+import static com.kyriba.conference.discount.domain.DiscountType.*;
+import static org.springframework.http.MediaType.APPLICATION_JSON_UTF8_VALUE;
+
+/**
+ * @author Igor Lizura
+ */
+@RestController
+@AllArgsConstructor
+@RequestMapping("/v1/discount")
```
Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: Use common pattern /api/v1/...

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.List;
+
+import static com.kyriba.conference.discount.domain.DiscountType.*;
+import static org.springframework.http.MediaType.APPLICATION_JSON_UTF8_VALUE;
+
+/**
+ * @author Igor Lizura
+ */
+@RestController
+@AllArgsConstructor
```
Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: Useless annotation here.

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+@RequestMapping("/v1/discount")
+public class DiscountController {
+    @ResponseBody
+    @GetMapping(produces = APPLICATION_JSON_UTF8_VALUE)
+    public ResponseEntity<List<DiscountDto>> getDiscounts() {
```
Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: Wrapping in response entity is useless here.

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+import static org.springframework.http.MediaType.APPLICATION_JSON_UTF8_VALUE;
+
+/**
+ * @author Igor Lizura
+ */
+@RestController
+@AllArgsConstructor
+@RequestMapping("/v1/discount")
+public class DiscountController {
+    @ResponseBody
```
Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: Useless annotation.

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    @GetMapping(produces = APPLICATION_JSON_UTF8_VALUE)
```
Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: Get discounts have to be on URL .../discounts

Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: produces and consumes may be configured at class level

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+    @ResponseBody
+    @PutMapping(value = "/{type}", produces = APPLICATION_JSON_UTF8_VALUE, consumes = APPLICATION_JSON_UTF8_VALUE)
+    public ResponseEntity<DiscountResponse> updateDiscount(@RequestBody DiscountDto discount) {
+        return new ResponseEntity<>(new DiscountResponse(STUDENT), HttpStatus.OK);
+    }
+
+    @ResponseBody
+    @DeleteMapping(value = "/{type}", produces = APPLICATION_JSON_UTF8_VALUE)
+    public ResponseEntity<?> deleteDiscount(@PathVariable String type) {
```
Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: There is no reason to return anything from delete method.

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+    @ResponseBody
+    @GetMapping(value = "/{type}", produces = APPLICATION_JSON_UTF8_VALUE)
+    public ResponseEntity<DiscountDto> getDiscount(@PathVariable String type) {
+        return new ResponseEntity<>(new DiscountDto(STUDENT, 30), HttpStatus.OK);
+    }
+
+    @ResponseBody
+    @PutMapping(value = "/{type}", produces = APPLICATION_JSON_UTF8_VALUE, consumes = APPLICATION_JSON_UTF8_VALUE)
+    public ResponseEntity<DiscountResponse> updateDiscount(@RequestBody DiscountDto discount) {
```
Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: The same issues as for other methods.

Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: public is not needed

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+    @ResponseBody
+    @PostMapping(produces = APPLICATION_JSON_UTF8_VALUE, consumes = APPLICATION_JSON_UTF8_VALUE)
+    public ResponseEntity<DiscountResponse> createDiscount(@RequestBody DiscountDto discount) {
+        return new ResponseEntity<>(new DiscountResponse(STUDENT), HttpStatus.CREATED);
+    }
+
+    @ResponseBody
+    @GetMapping(value = "/{type}", produces = APPLICATION_JSON_UTF8_VALUE)
+    public ResponseEntity<DiscountDto> getDiscount(@PathVariable String type) {
```
Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: The same issues as for other methods.

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    @GetMapping(produces = APPLICATION_JSON_UTF8_VALUE)
+    public ResponseEntity<List<DiscountDto>> getDiscounts() {
+        List<DiscountDto> response = Arrays.asList(new DiscountDto(JUNIOR, 30),
+                new DiscountDto(STUDENT, 50));
+        return new ResponseEntity<>(response, HttpStatus.OK);
+    }
+
+    @ResponseBody
+    @PostMapping(produces = APPLICATION_JSON_UTF8_VALUE, consumes = APPLICATION_JSON_UTF8_VALUE)
+    public ResponseEntity<DiscountResponse> createDiscount(@RequestBody DiscountDto discount) {
```
Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: The same issues as for other methods.

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+import static com.kyriba.conference.discount.domain.DiscountType.*;
+import static org.springframework.http.MediaType.APPLICATION_JSON_UTF8_VALUE;
+
+/**
+ * @author Igor Lizura
+ */
+@RestController
+@AllArgsConstructor
+@RequestMapping("/v1/discount")
+public class DiscountController {
```
Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java:26]: There is no data validation

Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java:26]: There is no API documentation in Swagger

File: discount/src/main/java/com/kyriba/conference/discount/domain/DiscountType.java [hunk 16]
```diff
@@ -0,0 +1,12 @@
+package com.kyriba.conference.discount.domain;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+
+/**
+ * @author Igor Lizura
+ */
+@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
```
Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/domain/DiscountType.java]: Useless annotation.

File: discount/src/main/java/com/kyriba/conference/discount/domain/dto/DiscountDto.java [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+import lombok.NoArgsConstructor;
+import lombok.NonNull;
+
+/**
+ * @author Igor Lizura
+ */
+@Getter
+@NoArgsConstructor
+@AllArgsConstructor
+@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
```
Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/domain/dto/DiscountDto.java]: Useless annotation

File: discount/src/main/java/com/kyriba/conference/discount/domain/dto/DiscountDto.java [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+public class DiscountDto {
+    @NonNull
+    private DiscountType type;
+
+    int percentage;
```
Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/domain/dto/DiscountDto.java]: Any validation rules?

File: discount/src/main/java/com/kyriba/conference/discount/domain/dto/DiscountDto.java [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+import lombok.NonNull;
+
+/**
+ * @author Igor Lizura
+ */
+@Getter
+@NoArgsConstructor
+@AllArgsConstructor
+@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
+public class DiscountDto {
```
Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/domain/dto/DiscountDto.java]: No API documentation

File: discount/src/test/java/com/kyriba/conference/discount/DiscountApplicationTests.java [hunk 20]
```diff
@@ -0,0 +1,15 @@
+package com.kyriba.conference.discount;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.junit4.SpringRunner;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+public class DiscountApplicationTests {
```
Comment (xpinjection) [inline discount/src/test/java/com/kyriba/conference/discount/DiscountApplicationTests.java]: Useless test

File: discount/src/test/java/com/kyriba/conference/discount/DiscountControllerTest.java [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+import static junit.framework.TestCase.assertEquals;
+import static org.springframework.http.MediaType.APPLICATION_JSON_UTF8_VALUE;
+import static org.springframework.restdocs.restassured3.RestAssuredRestDocumentation.document;
+import static org.springframework.restdocs.restassured3.RestAssuredRestDocumentation.documentationConfiguration;
+
+/**
+ * @author Igor Lizura
+ */
+@RunWith(SpringRunner.class)
+@AutoConfigureMockMvc
```
Comment (xpinjection) [inline discount/src/test/java/com/kyriba/conference/discount/DiscountControllerTest.java]: Useless annotation

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java (hunk 12 above)
Comment (xpinjection) [inline discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: Why DTO is passed? What if another discount type is passed there?

File: discount/build.gradle (hunk 3 above)
Comment (kyribamstraining) [reply discount/build.gradle]: done

File: discount/src/main/java/com/kyriba/conference/discount/DiscountApplication.java (hunk 4 above)
Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/DiscountApplication.java]: done

File: discount/src/main/java/com/kyriba/conference/discount/DiscountApplication.java (hunk 5 above)
Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/DiscountApplication.java]: done

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java (hunk 6 above)
Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: done

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java (hunk 7 above)
Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: done

File: discount/src/main/java/com/kyriba/conference/discount/domain/dto/DiscountDto.java (hunk 18 above)
Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/domain/dto/DiscountDto.java]: done

File: discount/src/main/java/com/kyriba/conference/discount/domain/dto/DiscountDto.java (hunk 19 above)
Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/domain/dto/DiscountDto.java]: done

File: discount/src/test/java/com/kyriba/conference/discount/DiscountApplicationTests.java (hunk 20 above)
Comment (kyribamstraining) [reply discount/src/test/java/com/kyriba/conference/discount/DiscountApplicationTests.java]: done

File: discount/src/test/java/com/kyriba/conference/discount/DiscountControllerTest.java (hunk 21 above)
Comment (kyribamstraining) [reply discount/src/test/java/com/kyriba/conference/discount/DiscountControllerTest.java]: done

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java (hunk 12 above)
Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: done

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java (hunk 8 above)
Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: done

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java (hunk 9 above)
Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: done

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java (hunk 10 above)
Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: done

Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: done

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java (hunk 11 above)
Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: done

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java (hunk 12 above)
Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: done

Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: done

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java (hunk 13 above)
Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: done

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java (hunk 14 above)
Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java]: done

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java (hunk 15 above)
Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java:26]: done

Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java:26]: done

File: discount/src/main/java/com/kyriba/conference/discount/domain/DiscountType.java (hunk 16 above)
Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/domain/DiscountType.java]: done

File: discount/src/main/java/com/kyriba/conference/discount/domain/dto/DiscountDto.java (hunk 17 above)
Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/domain/dto/DiscountDto.java]: done

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+    @DeleteMapping("/{type}")
+    void deleteDiscount(@ApiParam(value = "Discount type", required = true) @PathVariable String type) {
+    }
+
+    @ApiModel(value = "DiscountType", description = "Discount response")
+    @Value
+    private class DiscountResponse {
+        @ApiModelProperty(value = "Discount type")
+        @NotNull
+        DiscountType type;
```
Comment (ilya-a87) [inline discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java:64]: DiscountType is domain object exposed in API

File: discount/src/main/java/com/kyriba/conference/discount/api/dto/DiscountPercentageDto.java [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+
+import javax.validation.constraints.Max;
+import javax.validation.constraints.Min;
+
+/**
+ * @author Igor Lizura
+ */
+@ApiModel(value = "DiscountPercentage", description = "Discount updation request")
+@Getter
+public class DiscountPercentageDto {
```
Comment (ilya-a87) [inline discount/src/main/java/com/kyriba/conference/discount/api/dto/DiscountPercentageDto.java]: what is a difference between DiscountPercentageDto and DiscountDto? There is "percentage" in both.

File: discount/src/test/java/com/kyriba/conference/discount/DiscountControllerTest.java (hunk 21 above)
Comment (ilya-a87) [reply discount/src/test/java/com/kyriba/conference/discount/DiscountControllerTest.java]: @AutoConfigureMockMvc is not removed

File: discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java (hunk 22 above)
Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/api/DiscountController.java:64]: it's ok for me

File: discount/src/main/java/com/kyriba/conference/discount/api/dto/DiscountPercentageDto.java (hunk 23 above)
Comment (kyribamstraining) [reply discount/src/main/java/com/kyriba/conference/discount/api/dto/DiscountPercentageDto.java]: was renamed

File: discount/src/test/java/com/kyriba/conference/discount/DiscountControllerTest.java (hunk 21 above)
Comment (kyribamstraining) [reply discount/src/test/java/com/kyriba/conference/discount/DiscountControllerTest.java]: done

**Review by ilya-a87 — APPROVED** [review]

**Pull request merged**